### PR TITLE
Add Bucketing support to TarredAudioToClassificationLabelDataset

### PR DESCRIPTION
**Pull request by entn-at**
# What does this PR do ?

This PR adds support for bucketing to `TarredAudioToClassificationLabelDataset`.

**Collection**: [ASR]

# Changelog 
- Modify `_setup_dataloader_from_config` in `nemo.collections.asr.models.classification_models._EncDecBaseModel` to enable bucketing.
- Adds `pad_id` member to `TarredAudioToClassificationLabelDataset` class.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc) -> It does not
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries?

**PR Type**:
- [x] New Feature

### CODE REVIEW

**Review by titu1994 — COMMENTED** [review]
Thanks for the PR ! It overall looks correct, @VahidooX can you review ? I think there should be somewhere to use the path resolved for bucketing that's there in tarred datasets.

**Comment by entn-at** [comment]
Rebased on current main

File: tests/collections/asr/test_asr_classification_model.py [hunk 1]
```diff
@@ -191,6 +191,8 @@ def test_EncDecClassificationDatasetConfig_for_AudioToSpeechLabelDataset(self):
             'sample_rate',
             'normalize_audio',
             'augmentor',
+            'bucketing_batch_size',
```
Comment (VahidooX) [inline tests/collections/asr/test_asr_classification_model.py:194]: These new parameters need to be added to the config file in the nemo/collections/asr/models/configs/classification_models_config.py.

File: nemo/collections/asr/data/audio_to_label_dataset.py [hunk 2]
```diff
@@ -78,21 +78,41 @@ def get_tarred_classification_label_dataset(
...
-        is_regression_task=config.get('is_regression_task', False),
-    )
-    return dataset
+    tarred_audio_filepaths = config['tarred_audio_filepaths']
+    manifest_filepaths = config['manifest_filepath']
+    datasets = []
+    tarred_audio_filepaths = convert_to_config_list(tarred_audio_filepaths)
+    manifest_filepaths = convert_to_config_list(manifest_filepaths)
+
+    if len(manifest_filepaths) != len(tarred_audio_filepaths):
```
Comment (VahidooX) [inline nemo/collections/asr/data/audio_to_label_dataset.py:87]: Would you please add the support for weighting here?
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/data/audio_to_text_dataset.py#L165

here is a PR example which added weighting for another model:
https://github.com/NVIDIA/NeMo/pull/4530

File: tests/collections/asr/test_asr_classification_model.py (hunk 1 above)
Comment (VahidooX) [reply tests/collections/asr/test_asr_classification_model.py:194]: Also please add them to the config files in the example folder.

**Review by VahidooX — CHANGES_REQUESTED** [review]
LGTM! Just need to add weighting and also add the new params to the configs.

File: nemo/collections/asr/data/audio_to_label.py [hunk 3]
```diff
@@ -709,6 +709,40 @@ class TarredAudioToClassificationLabelDataset(_TarredAudioLabelDataset):
     # self.labels = labels if labels else self.collection.uniq_labels
     # self.num_commands = len(self.labels)
 
+    def __init__(
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/asr/data/audio_to_label.py:712]: Agreed, doesn't seem like this change is necessary.

[the remark this answers is not in the document]
Comment (entn-at) [reply nemo/collections/asr/data/audio_to_label.py]: I reverted this change.

File: tests/collections/asr/test_asr_classification_model.py (hunk 1 above)
Comment (entn-at) [reply tests/collections/asr/test_asr_classification_model.py:194]: Done

File: nemo/collections/asr/data/audio_to_label_dataset.py (hunk 2 above)
Comment (entn-at) [reply nemo/collections/asr/data/audio_to_label_dataset.py:93]: I added support for weighted bucketing

File: nemo/collections/asr/models/configs/classification_models_config.py [hunk 4]
```diff
@@ -53,6 +53,10 @@ class EncDecClassificationDatasetConfig(nemo.core.classes.dataset.DatasetConfig)
     normalize_audio: bool = False
     is_regression_task: bool = False
 
+    # bucketing params
+    bucketing_strategy: str = "synced_randomized"
```
Comment (titu1994) [inline nemo/collections/asr/models/configs/classification_models_config.py:57]: Can you add bucketing strategy and bucketing batch size in the configs as well ?

**Comment by titu1994** [comment]
Can you run the following at the root of your nemo repo - `python setup.py style --scope nemo` and then `python setup.py style --fix --scope nemo`.

File: nemo/collections/asr/models/configs/classification_models_config.py (hunk 4 above)
Comment (entn-at) [reply nemo/collections/asr/models/configs/classification_models_config.py:57]: `bucketing_strategy` and `bucketing_batch_size` were already present in the marblenet and matchboxnet example configs. According to git blame, they were added as part of the PR that added randomized bucketing (https://github.com/NVIDIA/NeMo/pull/3445).

**Comment by entn-at** [comment]
Thanks @titu1994, I ran `python setup.py style --fix --scope nemo` and added a commit with the style fixes.

File: nemo/collections/asr/models/configs/classification_models_config.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/models/configs/classification_models_config.py:57]: Perfect. Thanks

**Review by VahidooX — APPROVED** [review]
LGTM, thank you for the contribution!

**Pull request merged**